Include kinds in kind mismatch error message

curiousleo opened the pull request:
Closes https://github.com/ocaml/ocaml/issues/11529.

# User-visible changes

For the following code:

```ocaml
module M : sig type t end = struct type t = { i : int } end
type t = M.t = { i : int };;
```

**Before**

```
2 | type t = M.t = { i : int };;
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This variant or record definition does not match that of type M.t
       Their kinds differ.
```

**After**

```
2 | type t = M.t = { i : int };;
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This variant or record definition does not match that of type M.t
       The original is abstract, but this is a record.
```

Similarly when one type is a record and the other is a variant or extensible variant.

curiousleo committed:
Include kinds in "Their kinds differ" error

Changes the error message for code like the following:

    module M : sig type t end = struct type t = { i : int } end
    type t = M.t = { i : int };;

Before:

    2 | type t = M.t = { i : int };;
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
    Error: This variant or record definition does not match that of type M.t
           Their kinds differ.

After:

    2 | type t = M.t = { i : int };;
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
    Error: This variant or record definition does not match that of type M.t
           The original is abstract, but this is a record.

curiousleo commented on typing/includecore.ml:
```diff
@@ -174,10 +174,22 @@ type variant_change =
   (Types.constructor_declaration as 'l, 'l, constructor_mismatch)
     Diffing_with_keys.change
 
+type type_kind =
+  | Kind_abstract
+  | Kind_record
+  | Kind_variant
+  | Kind_open
+
+let of_kind = function
+  | Type_abstract -> Kind_abstract
+  | Type_record (_, _) -> Kind_record
+  | Type_variant (_, _) -> Kind_variant
+  | Type_open -> Kind_open
```
I introduced `type_kind` here to avoid having to put `('lbl, 'cstr) Types.type_kind` into the `Kind` constructor for `type_mismatch`. Using `type_kind` there instead of `('lbl, 'cstr) Types.type_kind` just seems simpler, but it does require introducing this additional `type_kind` type that's really only used in one place.

Octachron replied:
It sounds like you could use  `Types.type_decl_kind`?

Octachron replied:
At the same time, I agree that it is a good practice for the error data types to contain as little information as needed. 
Ideally, I would like all error data types to be serializable in a machine readable format.

curiousleo replied:
Good point, `Types.type_decl_kind` is a better candidate than `Types.type_kind`. But it seems like there is no strong objection to introducing an even simpler type in `Includecore`, so I'll go with that for now.

gasche commented on typing/includecore.ml:
```diff
@@ -386,8 +408,8 @@ let report_type_mismatch first second decl env ppf err =
-  | Kind ->
-      pr "Their kinds differ."
+  | Kind (kind1, kind2) ->
+      report_kind_mismatch ppf first second kind1 kind2
```
It would be more consistent with other branches to have

```ocaml
| Kind err ->
  report_kind_mismatch first second ppf err
```

which requires slightly different definitions (probably a `kind_mismatch` type being defined as `type_kind * type_kind`.)

curiousleo replied:
Done!

gasche reviewed (commented):
This generally looks like a nice improvement, thanks!

I don't see any issue with introducing a new specialized type, and in fact I suggest a second type definition in a style comment below.

Octachron commented on typing/includecore.ml:
```diff
@@ -378,6 +390,16 @@ let report_private_object_mismatch env ppf err =
+  | Kind_record -> "a record"
+  | Kind_variant -> "a variant"
+  | Kind_open -> "an extensible variant" in
+  pr "Their kinds differ: %s is %s, but %s is %s"
```
This is a good occasion to remove the never explained notion of `kind` from the error message.
```
This is a variant, but the original is an extensible variant
```
is as informative as
```
Their kind differ: this is a variant, but the original is an extensible variant
```

curiousleo replied:
Cool, I've removed "Their kinds differ"!

gasche commented:
I see no reason *not* to have this change, so I think we (@Octachron and myself) will approve and merge once the (minor) review feedback is taken into account.

curiousleo commented:
> I see no reason _not_ to have this change, so I think we (@Octachron and myself) will approve and merge once the (minor) review feedback is taken into account.

Awesome! I'll make the changes you and @Octachron suggested. Thanks for the incredibly fast review!

Octachron commented:
I agree that expanding the sibylline original error message is a clear improvement.

gasche commented:
It looks like the CI fails because more testsuite results should be updated. You can use "make promote" from the `testsuite/` directory for this -- but then you should review the `git diff` to ensure the changes are as you intended.

curiousleo commented:
> It looks like the CI fails because more testsuite results should be updated. You can use "make promote" from the `testsuite/` directory for this -- but then you should review the `git diff` to ensure the changes are as you intended.

Looks like it! I think I fixed them all now.

gasche commented on Changes:
```diff
@@ -84,6 +84,9 @@ Working version
 ### Compiler user-interface and warnings:
 
+- #11530: Include kinds in kind mismatch error message.
+  (Leonhard Markert, review by TODO)
```
TODO is now Florian and me

curiousleo replied:
... déjà vu!

https://github.com/ocaml/ocaml/blob/ef8431146cb028c9d09db30bd9a327ee079d920b/Changes#L37-L38

curiousleo commented on typing/includecore.ml:
```diff
@@ -378,6 +392,19 @@ let report_private_object_mismatch env ppf err =
+  | Kind_record -> "a record"
+  | Kind_variant -> "a variant"
+  | Kind_open -> "an extensible variant" in
+  pr "%s is %s, but %s is %s."
+    (String.capitalize_ascii second)
+    (kind_to_string kind2)
+    first
+    (kind_to_string kind1)
```
Regarding the order: I put `second` before `first` here because I thought that "This variant or record ... This is ..., but the original is ..." was perhaps clearer because in both cases "this" comes first, e.g.:

(`second` then `first`, the current implementation)

```
Error: This variant or record definition does not match that of type
         ('a, [> `A ]) def
       This is a variant, but the original is a record.
```

Versus `first` then `second`

```
Error: This variant or record definition does not match that of type
         ('a, [> `A ]) def
       The original is a record, but this is a variant.
```

However in other contexts, it seems strange:

(`second` then `first`, the current implementation)

```
        The second is a record, but the first is a variant.
```

Versus `first` then `second`

```
        The first is a variant, but the second is a record.
```

---

I'm inclined to switch the order to `first` then `second` the get rid of "The second is ..., but the first is ...". Any thoughts on this?

curiousleo replied:
I've switched the order to `first` then `second`.

Octachron replied:
I would suggest to keep the `first` then `second` order.
If we want to optimize the message order later, the right fix from my point of view is to compose the error message at the phrase level and avoid inserting words that accidentally fit in the grammatical context. 
This is a necessity for localization, but I think that this is a good guideline in general.

curiousleo commented:
Weird, it looks like the GitHub UI only lets me request a review from one reviewer at a time ... ? I've addressed all comments and believe that the test suite should now pass, ready for a second round of reviews!

gasche commented:
Merged, thanks!
